crates/core/src/kernel/snapshot/iterators/scan_row.rs

```diff
@@ -0,0 +1,455 @@
+use std::collections::HashMap;
+use std::pin::Pin;
+use std::sync::{Arc, LazyLock};
+use std::task::{Context, Poll};
+
+use arrow::array::{Array as _, *};
+use arrow_schema::{Field as ArrowField, Fields};
+use arrow_schema::{Field, Schema};
+use delta_kernel::engine::arrow_conversion::TryIntoArrow as _;
+use delta_kernel::engine::arrow_conversion::TryIntoKernel;
+use delta_kernel::engine::arrow_data::ArrowEngineData;
+use delta_kernel::engine::parse_json;
+use delta_kernel::expressions::Scalar;
+use delta_kernel::expressions::UnaryExpressionOp;
+use delta_kernel::scan::scan_row_schema;
+use delta_kernel::schema::DataType;
+use delta_kernel::schema::PrimitiveType;
+use delta_kernel::snapshot::Snapshot as KernelSnapshot;
+use delta_kernel::{EvaluationHandler, Expression, ExpressionEvaluator};
+use futures::Stream;
+use pin_project_lite::pin_project;
+use tracing::log::*;
+
+use crate::kernel::arrow::engine_ext::SnapshotExt;
+use crate::kernel::arrow::extract::{self as ex};
+use crate::kernel::StructType;
+use crate::kernel::ARROW_HANDLER;
+use crate::{DeltaResult, DeltaTableError};
+
+pin_project! {
+    pub(crate) struct ScanRowOutStream<S> {
+        snapshot: Arc<KernelSnapshot>,
+
+        #[pin]
+        stream: S,
+    }
+}
+
+impl<S> ScanRowOutStream<S> {
+    pub fn new(snapshot: Arc<KernelSnapshot>, stream: S) -> Self {
+        Self { snapshot, stream }
+    }
+}
+
+impl<S> Stream for ScanRowOutStream<S>
+where
+    S: Stream<Item = DeltaResult<RecordBatch>>,
+{
+    type Item = DeltaResult<RecordBatch>;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let this = self.project();
+        match this.stream.poll_next(cx) {
+            Poll::Ready(Some(Ok(batch))) => match parse_stats_column(&this.snapshot, &batch) {
+                Ok(batch) => Poll::Ready(Some(Ok(batch))),
+                Err(err) => Poll::Ready(Some(Err(err))),
+            },
+            other => other,
+        }
+    }
```

Comment:
it seems work has started to support async/streams directly from kernel. As such we start to move some processing onto streams rather that doing it in iterator world.

This should also align well when we work on the datafusion integrations, since there we find the same model.

Reply:
`[citation needed]` :laughing: 

There's been a lot of talk that I have heard but I haven't seen any concrete changes, do you have some to link?